src/peft/tuners/lora/eva.py

```diff
@@ -0,0 +1,463 @@
+        if k in equal_inputs_map:
+            continue
+        if isinstance(prepare_activations_fn, Mapping):
+            try:
+                fn = prepare_activations_fn[k]
+            except KeyError:
+                raise ValueError(f"if prepare_activations_fn is a mapping it must contain adapter module name {k}")
+        else:
+            fn = prepare_activations_fn
+        hooks[k] = SVDHook(k, max_components, peft_config.eva_config.tau, fn)
+    layer_hook_map = {**dict(zip(hooks.keys(), hooks.keys())), **equal_inputs_map}
+
+    # start svd calculation
+    pbar = tqdm(iter(cycle(dataloader)), position=0, leave=False)
+    convergence_dict = {k: False for k in hooks.keys()}
+    rank_dist = {k: max_components for k in layer_hook_map.keys()}
+    for inputs in pbar:
+
+        indices = None
+        if get_indices_fn is not None:
+            indices = get_indices_fn(inputs, peft_config)
+
+        for name, hook in hooks.items():
+            module = model.get_submodule(name)
+            module._forward_hooks.clear()
+            # check if all components that are needed for the rank distribution have converged
+            if torch.all(hook.converged[:rank_dist[name]]):
+                convergence_dict[name] = True
+                continue
+            convergence_dict[name] = False
+            hook.indices = indices.T.unbind()
+            module.register_forward_hook(hook)
+
+        layer_converged = list(convergence_dict.values()) + [convergence_dict[v] for v in equal_inputs_map.values()]
+        pbar.set_description(f"{sum(layer_converged)}/{len(layer_converged)} layers have converged")
+        
+        if all(convergence_dict.values()):
+            break
+
+        forward_fn(model, move_inputs_to_device(inputs, device))
+
+        # in case some hooks have to skip the svd calculation because the number of tokens is less than the number of components
+        if not all([hasattr(h.svd, "components_") for h in hooks.values()]):
+            continue
+
+        rank_dist = _get_rank_distribution(hooks, layer_hook_map, equal_inputs_map, rank_budget, max_components)
+
+    eva_state_dict = {}
+    for name, rank in rank_dist.items():
+        if rank == 0:
+            continue
+        hook = hooks[layer_hook_map[name]]
+        assert torch.all(hook.converged[:rank]) # this should never happen because we check for convergence
+        u = hook.svd.components_[:rank]
+        if peft_config.eva_config.whiten:
+            u /= hook.svd.singular_values_[:rank].sqrt().reshape(-1, 1)
+        eva_state_dict[name] = u
+
+    # objects are torch tensors on the model device
+    eva_state_dict = {k: v.to(device) for k, v in eva_state_dict.items()}
+
+    # restore model state
+    model.train(training)
+
+    return eva_state_dict
+
+
+@torch.no_grad()
+def initialize_lora_eva_weights(
+    model,
+    peft_config,
+    dataloader,
+    forward_fn: callable = lambda model, inputs: model(**inputs),
```

Comment:
Let's make this optional, and if `None` is passed, define a local function and use that for `forward_fn`. Let's avoid assigning a `lambda` function.